mlrun/utils/async_http.py

```diff
@@ -0,0 +1,264 @@
+# Copyright 2018 Iguazio
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import asyncio
+import http
+import typing
+from typing import List, Optional
+
+import aiohttp
+import aiohttp.http_exceptions
+from aiohttp_retry import ExponentialRetry, RequestParams, RetryClient, RetryOptionsBase
+from aiohttp_retry.client import _RequestContext
+
+from ..config import config
+from ..errors import err_to_str
+from .helpers import logger as mlrun_logger
+
+
+class AsyncClientWithRetry(RetryClient):
+    """
+    Extends by adding retry logic on both error statuses and certain exceptions.
+    """
+
+    def __init__(
+        self,
+        max_retries=config.http_retry_defaults.max_retries,
+        retry_backoff_factor=config.http_retry_defaults.backoff_factor,
+        retry_on_exception=True,
+        retry_on_status=True,
```

Comment:
tbh, the reason the name is wrong is because i have copied the logic from the sync client
which also gets `retry_on_status` but pass `raise_for_status=retry_on_status` later on. I will update on the async client to use `retry_on_status` for better reflection